[Merged by Bors] - feat(analysis/normed/group/add_circle): define the additive circle

ocfnash opened the pull request:
---

- [x] depends on: #16130
- [x] depends on: #16158
- [x] depends on: #16135


[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(algebra/order/floor): lemmas about `int.floor`, `int.ceil`, `int.fract`, `round`

ocfnash committed:
feat(analysis/normed/group/add_circle): define the additive circle

ocfnash committed:
Suggestions from CR

ocfnash committed:
Add lemma `int.ceil_coe'`

ocfnash committed:
CR suggestion: add lemmas `int.floor_le_sub_one_iff`, `int.add_one_le_ceil_iff`

ocfnash committed:
Update src/algebra/order/floor.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

ocfnash committed:
Update naming: change `_coe`, `_coe'` suffixes to `_int_cast`, `_nat_cast`.

ocfnash committed:
Update naming 2: change `_coe`, `_coe'` suffixes to `_int_cast`, `_nat_cast`.

The sequel

ocfnash committed:
Fix long lines

jsm28 commented on src/analysis/normed/group/add_circle.lean:
```diff
@@ -0,0 +1,234 @@
+      ext,
+      simp [(by { rw hz, abel, } : y = z • p + x), mul_div_cancel _ hp.ne.symm, add_div], },
+    exact quotient.lift _ this,
+  end,
```
This feels like it's much more complicated than necessary (compared to the definition of `real.angle.to_real`). I'd expect something like `λ x, ⟨(to_Ico_mod_periodic 0 hp).lift x, to_Ico_mod_mem_Ico 0 hp x⟩` to suffice (and maybe some other proofs might also be simpler, considering those for `to_real`).

jsm28 replied:
(Maybe `to_Ico_mod_mem_Ico` doesn't have quite the right type, because it actually asserts membership of `Ico 0 (0 + p)`, but correcting for that should be straightforward.)

ocfnash replied:
Thanks, yes this is just a silly proof I bashed out so I could push something sorry-free.

I hope to return to this PR in a day or two and I expect using `to_Ico_mod_mem_Ico` for inspiration here will help. Thanks again.

ocfnash committed:
Merge branch 'master' into ocfnash/floor_rings

ocfnash committed:
Fix after merging master

The relevant changes arrived in master via https://github.com/leanprover-community/mathlib/pull/16025

ocfnash committed:
Merge branch 'master' into ocfnash/add_circle

ocfnash committed:
Golfing and linting

ocfnash committed:
Merge branch 'master' into ocfnash/add_circle

jsm28 commented on src/algebra/order/archimedean.lean:
```diff
@@ -331,7 +331,8 @@ floor_ring.of_floor α (λ a, classical.some (exists_floor a))
 
 /-- A linear ordered field that is a floor ring is archimedean. -/
-lemma floor_ring.archimedean (α) [linear_ordered_field α] [floor_ring α] : archimedean α :=
+@[priority 100] -- see Note [lower instance priority]
+instance floor_ring.archimedean (α) [linear_ordered_field α] [floor_ring α] : archimedean α :=
```
Given this change, the `local attribute [instance] floor_ring.archimedean` in `to_interval_mod.lean` should no longer be needed.

ocfnash replied:
Good point, done.

ocfnash committed:
Drop redundant local attribute

hrmacbeth commented on src/analysis/normed/group/add_circle.lean:
```diff
@@ -0,0 +1,213 @@
+## TODO
+
+ * Link with periodicity
+ * Measure space structure
```
See the branch `RmodZ` for the measure space structure, particularly the theorem in `measure_theory/integral/periodic` that the quotient map from `ℝ` to `add_circle` is a measure-preserving map with respect to `volume.restrict (Ioc t (t + T))` on `ℝ` and the Haar measure on `add_circle`.

ocfnash replied:
Just had a look, thanks. It certainly looks useful!

hrmacbeth commented on src/analysis/normed/group/add_circle.lean:
```diff
@@ -0,0 +1,213 @@
+
+/-- An auxiliary definition used only for constructing `add_circle.equiv_add_circle`. -/
+private def equiv_add_circle_aux (hp : p ≠ 0) : add_circle p →+ add_circle q :=
+quotient_add_group.lift _
```
Is there a way to use [function.periodic.lift](https://leanprover-community.github.io/mathlib_docs/algebra/periodic.html#function.periodic.lift) for this?  (Untested.)

ocfnash replied:
It can certainly be used but at least my attempt to use it looks less promising than what I have already put in place:
```lean
private def equiv_add_circle_aux (hp : p ≠ 0) : add_circle p →+ add_circle q :=
have h : function.periodic (λ (x : 𝕜), (↑(q * (p⁻¹ * x)) : add_circle q)) p := λ x, by simp [mul_add, hp],
{ to_fun := h.lift,
  map_zero' := by { erw h.lift_coe 0, simp, },
  map_add' := sorry, }
```

hrmacbeth committed:
split into two files

hrmacbeth committed:
adjust title

jcommelin reviewed (approved):
Thanks :tada:

bors merge
